### PR TITLE
GS: Check memory overlaps with targets for GS downloads (Texture cache)

### LINKED ISSUE #3437: Shadow Hearts: Covenant  - Judgement Ring In HW Rendering
***PCSX2 version:***
1.6.0

***PCSX2 options:***
I've tried all EE/IOP and VU options, OPH Flag Hack, and all HW video hacks. Speedhacks on and off. No change.

***Plugins used:***
GL HW and D3D HW

***Description of the issue:***
I think this is fairly well-known, but I didn't see any existing issues for it.

No issues in software renderers, but in hardware renderers, the judgement ring (the timing-based input part of attacking in combat) will make the entire background turn black while it's on screen. What it's _supposed_ to do is to freeze frame the scene (you can tell that it's basically taking a screenshot because the resolution is lower than normal with upscaling). In prior versions of PCSX2 it used to make about half the screen black, and half of it frozen, in more recent versions it makes the entire background black. The particularly strange part is that it doesn't happen every time, sometimes it does exactly what it's supposed to do. You can use a savestate, start an attack, and sometimes you'll get black, load the state, wait a different amount of time before attacking, and sometimes it'll display correctly. Then, when you attack with a different character, sometimes it'll be black and sometimes it'll show a frozen frame from when the _previous_ character attacked.

The same attack is in both frames below. Hardware rendering in both. It worked correctly, loaded the state, displayed black. Sometimes it'll work, sometimes it won't.

image

image

***How to reproduce the issue:***
Use either hardware renderer, get into combat, and start an attack.

***Last known version to work:***
None

***PC specifications:***
i7 4770k / 32GB RAM / GTX 780 / W7 64

### PR DESCRIPTION
**Pull request by Fothsid**
### Description of Changes
Invalidates local memory for render targets whose memory regions overlap with the requested one, instead of just checking for the same TBP.

### Rationale behind Changes
Games like Shadow Hearts Covenant offset their GS downloads by specifying source Y coordinate that is beyond the texture size. Without this change, the game experiences semi-random rendering issues when displaying the Judgement ring in battle, as detailed in #3437. This PR essentially fixes that issue. 

### Suggested Testing Steps
While this PR fixes #3437, it could easily introduce new issues to other games, which is something that needs to be tested. For Shadow Hearts Covenant, enter a battle and try to attack/use skill or item, the game shouldn't display any black screen or wrong shots when the Judgement ring appears.

### CODE REVIEW

File: pcsx2/GS/Renderers/HW/GSTextureCache.cpp [hunk 1]
```diff
@@ -1066,8 +1066,27 @@ void GSTextureCache::InvalidateLocalMem(const GSOffset& off, const GSVector4i& r
...
+			// using overly big source Y position values.
+			// Checking for targets that overlap with the requested memory region
+			// instead of just comparing TBPs should fix that.
+			// For example, this fixes Judgement ring rendering in Shadow Hearts 2.
+			GSVector4i gsmem_range(r.left, bp + r.top * bw, r.right, bp + r.bottom * bw);
+			GSVector4i tmem_range(t->m_valid.left,
+									t->m_TEX0.TBP0 + t->m_valid.top * t->m_TEX0.TBW,
+									t->m_valid.right,
+									t->m_TEX0.TBP0 + t->m_valid.bottom * t->m_TEX0.TBW);
+			GSVector4i intersection = gsmem_range.rintersect(tmem_range);
```
[the remark this answers is not in the document]
Comment (Fothsid) [reply pcsx2/GS/Renderers/HW/GSTextureCache.cpp]: Changed.

**Comment by refractionpcsx2** [comment]
Sorry, I have bad news, breaks the dirt on jake's legs in Dog's Life (can't GS Dump it either cos it does it as you play)

I can find out what values it's expecting for you, if you want?

image

File: pcsx2/GS/Renderers/HW/GSTextureCache.cpp [hunk 2]
```diff
@@ -1066,8 +1066,27 @@ void GSTextureCache::InvalidateLocalMem(const GSOffset& off, const GSVector4i& r
...
+			// instead of just comparing TBPs should fix that.
+			// For example, this fixes Judgement ring rendering in Shadow Hearts 2.
+			const GSVector4i gsmem_range(r.left, bp + r.top * bw, r.right, bp + r.bottom * bw);
+			const GSVector4i tmem_range(t->m_valid.left,
+									t->m_TEX0.TBP0 + t->m_valid.top * t->m_TEX0.TBW,
+									t->m_valid.right,
+									t->m_TEX0.TBP0 + t->m_valid.bottom * t->m_TEX0.TBW);
+			const GSVector4i intersection = gsmem_range.rintersect(tmem_range);
+			
+			if (intersection.width() && intersection.height() && GSUtil::HasSharedBits(psm, t->m_TEX0.PSM))
```
Comment (refractionpcsx2) [inline pcsx2/GS/Renderers/HW/GSTextureCache.cpp:1081]: instead of doing this you could do `!intersection.rempty()`

**Comment by Fothsid** [comment]
Can you log what values it passes to the InvalidateLocalMem function?

**Comment by refractionpcsx2** [comment]
it seems to be doing this

```[   44.2445] Read FIFO SBP 0x2a00 SBW 0x1 sx 0 sy 0 w 64 h 16 SPSM 0x0 bpp 32
[   44.2541] Read FIFO SBP 0x3de0 SBW 0x2 sx 0 sy 0 w 128 h 128 SPSM 0x2 bpp 16```

Now it's possible you've just poked it enough to break it, I know this is kinda janky anyway (my SW CLUT PR kinda fixes it better, so might behave different there? I could try tacking the two together)

**Comment by Fothsid** [comment]
Huh, doesn't look like they're doing anything funky there. Except, it does 16bpp for the second one, I wonder whether that could be breaking it.

**Comment by refractionpcsx2** [comment]
okay it doesn't work with my transplant either, though it breaks less, probably because there's nothing to invalidate

you have to admire his lovely pants tho

image

**Comment by Fothsid** [comment]
@refractionpcsx2 Could you check whether the last commit fixes the issue when combined with your SW CLUT PR?

**Review by JordanTheToaster — APPROVED** [review]

**Pull request merged**